### ISSUE TITLE
RFC's differ for Media Types depending on context.

### ISSUE THREAD

**Issue by rustonaut** [hyperium/mime#73]
Depending on where the Media Type is used it can have different grammar.

The main difference is between Media Types wrt. HTTP and Media Types wrt. MIME,
for the later on there are also additional aspects, like if it does appear in an internationalized
mail.

So we need at last a way to parametrize the parser, maybe more.

---

## Differences:

-  HTTP uses linear whitespaces which are either `' '` or `'\t'` but MIME use FWS which can contain soft line breaks (`"\r\n "` and `"\r\n\t"`)
-  "token" in HTTP Media Types _can not_ contain `{`, `}` but "token" in MIME Media Types _can_,
- tokens registered with IANA a further constraint (rfc6838)
  For HTTP this can be ignored but MIME defines type/subtype to either be `iana-token`, `ietf-token` or `x-token`. Both `iana-token` and `ietf-token` are tokens registered with IANA and therefore are restricted by rfc6838.  Also the restrictions in rfc6838 do also apply to parameter for registered Media Types, which also means they do not apply if either type or subtype are `x-tokens`

- additionally rfc6532 introduces internationalized mails and extends `quoted-string` to be able to contain all non-us-ascii-UTF-8 character, which HTTP does not allow (excluding obsolete syntax for now) but only in internationalized mails
- then there is rfc2231 which allows the encoding of parameter of Media Types in MIME (not HTTP) based on any registered charset. And adds support for specifying the language which is meant to be usefull for screen readers and allows spliting parameters into multiple parts to compline with Mail line length limits.

- [EDIT] in MIME Content-Type headers Media-Types can contain comments which the parser should skip over, for HTTP/1.1 this is malformed 

And that is if we ignore the obsolete syntax if we include it then

- in HTTP quoted-strings can contain any char 0x00-0xFF which can be anything including invalid or malformed utf8 (i.e. it can be not representable in rust strings as it add basically arbitrary bytes)
- in MIME quoted-strings _additionally_ contain any CTL (including `DEL`/`0x7f`, except `'\t'`, `'\n'`, `'\r'` through they could already be contained before, that doesn't change)

**Comment by dathinab** [comment]
My current plan would be to:
1. parametrize the parser
2. parse into a `MediaType` type which is comparable to the current `Mime` type
3. have methods with preset parameters which return a wrapper type containing `MediaType` to "keep" the information about e.g. it being HTTP/1.1 compatible in the type (need to scratch it out a bit more)
4. make the parse method return `StrictMediaType` which is a restrictive media type usable on all places expecting Media Types, which mainly means it would be rfc6838 compliant and ascii only. Which btw. means it is compatible with any IANA registered media type and should be compatible with most other ones too, (I _guess_ it would be >90% of all media types in practice)

**Comment by michaelbeaumont** [comment]
Hi, I just want to point out [RFC2387](https://tools.ietf.org/html/rfc2387). It allows specifying a mime type within a parameter (e.g. `type=application/json`) on which the current code chokes because `/` isn't a token.

**Comment by dathinab** [comment]
@michaelbeaumont thanks, I wasn't aware of this RFC.

I just skimmed over it and it seems that it does not extend the syntax of token, i.e. if you have 
a media type as value you have to quote it. E.g. `type="application/json"` which the parser _should_
already support (I haven't checked if it does as I'm currently rewriting the parser anyway)

**Comment by seanmonstar** [comment]
Ah yes, it's always fun trying to find which exact RFC contains a certain part of the rules!

What I wonder is if there is a good reason to actually have the implementation differ here, or if it's just better to allow the rules from both. For instance, while the HTTP RFCs may mention some things as being illegal, if browsers and other implementations allow them, perhaps we should too.

**Comment by dathinab** [comment]
The problem is that media types are used not just by browsers but also e.g. Mail (or more correct MIME) where they in some points noticeably differ.

If you have one parser which allows all accepting a Media-Type as input of a function would no longer guarantee anything about it's wellformed ness and there is quite a lot which can cause troubles.

For my use case it currently means that I have to write my own parser anyway to validate if the mime
for correctness, after the mime crates parser already did so, just not with all requirements.

The idea is to have a `AnyMediaType` which just contains the media type data without any information about how it was parsed into it.

Then use wrapper types  which provide parsing methods and dereference to `AnyMediaType` (+ Into<..> etc.) this allow libraries to require a MediaType to comply with a specific Specification but also allows them to not care at all.

Parts of the parser can be shared between all specifications using monomorphisation.
Monomorphisation and  dead code elimination prevents any overhead from appearing at runntime for not used Specifications, feature flags prevent compiler time overhead for not used Specifications.

I don't see any big drawbacks for this. Well except maybe that the code base increases slightly, but then mime seems to already have become the defacto crate for  any Media-Type usage so I thinks it's worth it.

**Comment by michaelbeaumont** [comment]
Somehow I missed your answer here! Good point  @dathinab, I suppose I would advocate for a lenient parser then, given that I'm parsing content types in API responses over which I have no control, and they are unfortunately unquoted.

**Comment by dathinab** [comment]
Additional to the standard conform parsing I plan to add a strict parser only parsing media types which could have been registered with IANA and a lenient parser which will parse some thinks without quoting _and adds quotes_.  ( I should be able to reuse most of the code for this from other parsers so it should not be to hard)

**Comment by dathinab** [comment]
TL;DR: I have a new parser(s) which does thinks "more correct" but this also means it does more work and does not implement PartialEq/Eq/Hash for MediaTypes (but Name/Value) so I'm not sure if this is ok for the mime crate which was mainly written for hyper.

Update:

I have a new parser which is generic over the exact mime grammar
spec which is used (http, mime, internationalized?, obsolte syntax?).
And it works well in general (I still have to add back in some of the optimizations used).

In the Media type specification the semantic interpretation of parameter values
_depends on the specification for the type/subtype_ which means:

- values could be case sensitivity but also could be case insensitive
- values might be equal even if not in a text comparison (e.g. alias of charsets like `utf-8`==`utf8`)
- values can be more complex structs, e.g. a codecs parameter with a list of codecs (which in them self might have aliases)

Also parameters have default values:

- e.g. `'text/plain; charset=us-ascii'` == `text/plain`

**Because of this implementing a general version of Eq/Hash or even matching media type ranges (in some cases) can not be implemented in a _general_ way.**

Through you can do special cases for some types/parameters, but this is tricky:
- people might expect media type matching/eq to "just work" if it's implemented  provoking bugs for any parameter not specially supported
- even for charset this would imply including information over all registered charsets aliases, braking with any non registered charset alias
- there are 1693 registered Media Types, which to socialize for??


Additionally to this implementing Eq/Hash correctly even if you "just compare the string representation" introduces performance penalties:

- parameters are order independent so you have to sort them either when parsing
  (doable but a slowdown) or every time you hash it _and_ to some degree when comparing (bad idea, hash calls should be kind of fast)



Because of this I came to the conclusion that it might be best not to implement Eq/Hash for Media Type, but instead let the user of the library handle this. And maybe provide some special helpers for very common cases.

While this sounds quite bad I do not think that outside of micro-examples/benchmarks etc. this is really a problem. Because any code which which uses media types (or media ranges) has to get and "evaluate" all parameter anyway.

E.g. some text software handling `text/html; charset=utf8` is unlikely to check if the media type is exact that it instead it would:

1. check if its start with the text type 
2. get the subtype check if it's supported 
3. go through the parameters stumble over charset
4. take the charset (utf8) as a string an pass it the the part of the application handling codecs to see if it's supported

Or instead of 1./2. it might be the case that they get them and see if they are in a set of supported types OR use them as a key for a hashmap to lookup a handle to continue processing (which is still doable as `(mtp.type_(), mtp.subtype()`)

---

While I do think this is an reasonable approach for most applications I'm not 100% sure if that's the case for Hyper (the structs it's used in derive PartialEq and I'm not sure if it's "just" for testing).



---
Lastly there is another think I stubled about which I thought was specific for Mime/Mail, that is you can percent encode parameter values by adding a `*` to the end of the parameter name e.g. `foo*=us-ascii''two%20words`.

I originally though it's just a think of Mime and does not nessesary needs to be put into a general media type implementation but sadly I was wrong (in mime it's still a bit more horrible by allowing parameters to be split into multiple parts). As a side note this can add a language tag to the parameter e.g. `us-ascii'de-De'blabla`.

The relevant rfcs are: rfc2231 (mime) and rfc8187 (http).

**Comment by seanmonstar** [comment]
Thanks for reading through and putting together this information from the various RFCs! Let me see if I can write up that I think the priorities of `mime` should be:

1. Correctness in the most common HTTP instances. Things in my mind like `text/plain`, `text/html; charset=utf8`, `application/json`, `image/png`, etc.
2. Fast
3. Easy-to-use.
4. Correct in edge-case or less common protocols than HTTP.

If it's possible to be more correct about some things, as long as it doesn't sacrifice 1, 2, or 3, then I'm all for it! But, if supporting certain media types that are proportionally barely used makes `mime` noticeably slower, or harder to use, in the more common cases, then I'd have to suggest that it doesn't belong.

Do those priorities seem reasonable for a library designed to be used for people working with HTTP?

**Comment by dathinab** [comment]
yes, very reasonable ;=)

The main open question is if it needs a Eq impl.

It's always possible to fall back to a less semantic Eq impl, i.e. comparing parameters case-sensitive by string value.

I just fear that this is likely to provoke bugs by people relying on Eq to  "just work".

**Comment by dathinab** [comment]
There is another thing which should be considered:

It might be a good idea to change subtype to return the full subtype,
but add some methods like e.g. `split_suffix` or similar.

Why so?
- the suffix is part of the registered subtype i.e. `svg+xml` is registered `svg` _on itself is not_. 
  (it's similar to the vendor tree of media types)
- in the rfc the subtype of `image/svg+xml` is `svg+xml`
- it's not part of the grammar, so there is no syntactical protection for it to make sense. E.g. `text/x.c++` is a valid media type but what is it's suffix (either it does not have one or it's an empty suffix). 
- all use case I'm currently aware of will use the full subtype including the suffix, treating `something+xml` as something complete different then `something+json` (through they "might"  share code internally). 

PS:
the vendor tree is how registered media types are structured, e.g. `application/vnd.adobe.flash.movie`

PPS:
There is also the think that the rfc states you should not add suffixes to registered media types (without registering them). So even if you have a svg represented as msgpack using `image/svg+msgpack` would be discouraged (you could use `image/x.svg+msgpack` instead). Which **I** interpret as specific handling of suffix not being intended to be a thing, well maybe it was intended once but if than it's no more.

**Comment by seanmonstar** [comment]
I suppose `Eq` and `Hash` could perhaps be removed. But reading the description of `Eq`:

> This means, that in addition to `a == b` and `a != b` being strict inverses, the equality must be (for all `a`, `b` and `c`):
>
> - reflexive: `a == a`;
> - symmetric: `a == b` implies `b == a`; and
> - transitive: `a == b` and `b == c` implies `a == c`.

It seems like this is actually true of the current `Mime` type.

----

It could be that `subtype()` is wrong, indeed. At least, according to the spec. What would users expect? Take `text/html+xml`. Would use a expect `subtype() == HTML` or `subtype() == HTML_XML`?

**Comment by dathinab** [comment]
> I suppose Eq and Hash could perhaps be removed. But reading the description of  [...]

Yes it's true one "syntax"/"raw value" level Eq+Hash can be implemented  (it's the fallback if it's needed, I also implemented it first and then noticed the problem with Eq/Hash so it's easy to use).

The problem is to get the semantics right e.g.::
- syntax level eq: `"text/x.my; a=b; d=e"` !=  `text/x.my; d=e; a=b`
- raw value level eq: `"text/x.my; a=b; d=e"` ==  `text/x.my; d=e; a=b`
 _but_ `text/plain; charset=us-ascii` != `text/plain` != `text/plain; charset=IBM367`
_and_ `text/plain; charset=UTF-8` != `text/plain; charset=utf-8`
- semantic correct eq: `text/plain; charset=us-ascii` == `text/plain` == `text/plain; charset=IBM367`

Currently it's done one a "raw value level eq" except for charset where it adds case-insensitive comparison and know specially about `utf8` to some degree.

I fare if Eq is provided people will not be aware about the fact that it's on a raw value level and not a semantic level and will introduce hard to find bugs in there software this way (basically the moment you use `==` on media types outside a test case it's super likely to be a bug, even more so because "unknown" params should be ignored so a software handling `text/plain` should also handle `text/plain; foo=bar`...)

---

I think I would expect `HTML_XML` because if I check if I can handle a media type and see it's HTML, I would probably go "yay can do" and then stumble over why this is not quite HTML, ~through in a certain sense HTML is special as it's one of the view cases where there is a suffix-less and a suffix version, and because HTML feels already quote close/similar to XML.~ wups turns out `text/html+xml` is not a registered media type (the correct one is `application/xhtml+xml`), through you should probably still be able to handle it.

Independent of this not having the suffix separated makes it harder to accidentally forget to check if the suffic is None/a expected suffix. E.g. if you get `image/png+bad` it super easy to forget to check if the suffix is None and try to handle it as `png` even through it's not.

**Comment by dathinab** [comment]
Just had a nice idea:
_why not feature gate eq+hash? (and document the problems with it)_

Pro:
- people will not "accidentally" use it
- it's still usable if needed/wanted e.g. for testing
- people are likely to read the doc why it's feature gated

Con:
- if they use it with a library which uses it with the feature enabled, due to the additivity of
  features they still can accidentally use Eq without even knowing that it's feature gated.
  (but this is more a limitation of rust features and might be solved in the future)

**Comment by SimonSapin** [comment]
https://github.com/hyperium/mime/issues/78 might be related, as it’s about an alternative parsing algorithm (https://mimesniff.spec.whatwg.org/#parsing-a-mime-type)

**Comment by dathinab** [comment]
thanks

**Comment by seanmonstar** [comment]
I've pushed to master two changes outlined here:

- Removed `Eq, Hash, PartialOrd, Ord` from `MediaType`/`MediaRange`.
- Fixed `subtype()` to include any +suffix. This also changes `mime::SVG` to be literally `"svg+xml"`.

I believe most other things have been taken care of in other PRs, so I'm going to close. If I missed something, we can make a more specific issue!
